File: inst/shiny/www/tlg_add_picker.js
// Client-side behaviour for the "Add TLGs to order" modal (see
// inst/shiny/functions/tlg_add_picker.R).  The modal shows dataset tabs
// (PK Concentrations / PK Parameters) over one column per output Type; this
// object drives search, tab switching, and the select-all helpers.  The
// per-open initialisation (active tab + first render) is injected inline by the
// R helper because it depends on the datasets present in the current modal.
window.tlgAdd = {
  q: "",
  tab: null,

  // visibleOnly: only toggle rows shown under the active tab / search, so
  // Select-all respects the current dataset.  Clear-all passes false to wipe
  // every dataset.
  _setChecked: function(scope, checked, visibleOnly) {
    var groups = {};
    scope.querySelectorAll(".checkbox").forEach(function(row) {
      if (visibleOnly && row.style.display === "none") return;
      var cb = row.querySelector("input[type=checkbox]");
      if (!cb) return;
      cb.checked = checked;
      var g = cb.closest(".shiny-input-checkboxgroup");
      if (g) groups[g.id] = g;
    });
    Object.keys(groups).forEach(function(id) {
      var inp = groups[id].querySelector("input[type=checkbox]");
      if (inp) inp.dispatchEvent(new Event("change", { bubbles: true }));
    });
  },

  selectAll: function() {
    var r = document.querySelector(".tlg-add-checklist");
    if (r) this._setChecked(r, true, true);
  },

  clearAll: function() {
    var r = document.querySelector(".tlg-add-checklist");
    if (r) this._setChecked(r, false, false);
  },

  colSelect: function(btn) {
    var c = btn.closest(".tlg-col");
    if (c) this._setChecked(c, true, true);
  },

  setQuery: function(v) {
    this.q = v || "";
    this.render();
  },

  setTab: function(v, btn) {
    this.tab = v;
    document.querySelectorAll(".tlg-tabs .tlg-tab").forEach(function(b) {
      b.classList.remove("active");
    });
    if (btn) btn.classList.add("active");
    this.render();
  },

  render: function() {
    var root = document.querySelector(".tlg-add-checklist");
    if (!root) return;
    var q = this.q.trim().toLowerCase();
    var tab = this.tab;
    // Per-dataset text-match counts are computed independently of the active
    // tab (they drive the tab badges), but only the active tab's matches are
    // shown.  This way searching in a term that only matches another dataset
    // surfaces the count on that tab's badge -- clicking it reveals the match
    // -- instead of silently showing another dataset's rows under this tab.
    var dsMatch = {};
    var totalMatches = 0;
    root.querySelectorAll(".tlg-ds").forEach(function(ds) {
      var name = ds.getAttribute("data-dataset");
      var isActive = name === tab;
      var anyVisible = false;
      ds.querySelectorAll(".checkbox").forEach(function(it) {
        var textMatch = q === "" || it.textContent.toLowerCase().indexOf(q) > -1;
        if (textMatch) {
          dsMatch[name] = (dsMatch[name] || 0) + 1;
          totalMatches++;
        }
        var show = isActive && textMatch;
        it.style.display = show ? "" : "none";
        if (show) anyVisible = true;
      });
      ds.style.display = anyVisible ? "" : "none";
    });
    // Per-column: visible count, empty state, hide select-all when empty.
    root.querySelectorAll(".tlg-col").forEach(function(col) {
      var vis = 0;
      col.querySelectorAll(".checkbox").forEach(function(it) {
        if (it.style.display !== "none") vis++;
      });
      var cnt = col.querySelector(".tlg-col-count");
      if (cnt) cnt.textContent = vis;
      var empty = col.querySelector(".tlg-col-empty");
      if (empty) empty.style.display = vis ? "none" : "";
      var sa = col.querySelector(".tlg-col-selall");
      if (sa) sa.style.display = vis ? "" : "none";
    });
    // Tab badges: per-dataset match count while searching, dataset total otherwise.
    document.querySelectorAll(".tlg-tabs .tlg-tab").forEach(function(t) {
      var badge = t.querySelector(".tlg-tab-count");
      if (!badge) return;
      badge.textContent = q === ""
        ? t.getAttribute("data-total")
        : (dsMatch[t.getAttribute("data-dataset")] || 0);
    });
    // Global "no matches" message only when nothing matches in any dataset.
    var nm = document.querySelector(".tlg-no-matches");
    if (nm) nm.style.display = totalMatches === 0 ? "" : "none";
  }
};

File: inst/shiny/functions/tlg_add_picker.R
#' Helpers for the "Add TLGs to order" modal (issue #1335).
#'
#' The picker was originally a grouped reactable whose Type/Dataset columns were
#' blank on every selectable row.  It is rebuilt as a catalog-style checklist:
#' dataset tabs (PK Concentrations / PK Parameters) over one column per output
#' Type (Tables / Listings / Graphs), with a search + download + select-all
#' toolbar, per-column select-all, and a live count on the confirm button.
#'
#' Client-side behaviour lives in `inst/shiny/www/tlg_add_picker.js`
#' (`window.tlgAdd`); styling lives in
#' `inst/shiny/www/styles/partials/_tlg_add_modal.scss`.  These helpers only
#' build the server-side UI and translate the checked rows back to `tlg_order()`
#' ids, keeping `tab_tlg_server()` small enough to stay under the cyclomatic
#' complexity limit.

# Fixed left-to-right order + icon for the Type columns, and dataset tab order.
.TLG_TYPE_ORDER <- c("Table", "Listing", "Graph")
.TLG_TYPE_ICON  <- c(Table = "table", Listing = "list-ul", Graph = "chart-line")
.TLG_DATASET_ORDER <- c("PK Concentrations", "PK Parameters")

#' Escape a string for safe embedding inside a single-quoted JS literal.
#' @param x Character vector.
#' @returns Character vector wrapped in single quotes with `'` and `\` escaped.
#' @noRd
tlg_js_str <- function(x) paste0("'", gsub("(['\\\\])", "\\\\\\1", x), "'")

#' Build the "Add TLGs" catalog checklist UI.
#'
#' @param avail Tibble of not-yet-selected TLGs (rows of `tlg_order()` with
#'   `Selection == FALSE`); must contain `Type`, `Dataset`, `Description`,
#'   `Link`, and `id`.
#' @param ns Namespace function for the calling module (`session$ns`).
#' @returns A list with `ui` (the modal body tag) and `group_ids` (the
#'   `checkboxGroupInput` ids created, read back on confirm).
#' @noRd
build_add_checklist <- function(avail, ns) {
  present_types <- intersect(.TLG_TYPE_ORDER, unique(avail$Type))
  datasets <- c(intersect(.TLG_DATASET_ORDER, unique(avail$Dataset)),
                setdiff(unique(avail$Dataset), .TLG_DATASET_ORDER))

  pairs <- dplyr::distinct(avail, Type, Dataset)
  pairs <- pairs[order(match(pairs$Type, .TLG_TYPE_ORDER), pairs$Dataset), ]
  pairs$input_id <- paste0("modal_check_", seq_len(nrow(pairs)))

  spec_icon_html <- as.character(icon("circle-info"))

  # One dataset block for a Type column: tagged with data-dataset so the tab bar
  # can show/hide it.  The active tab is applied client-side.
  build_group_ui <- function(type, dataset, input_id) {
    rows <- dplyr::filter(avail, Type == !!type, Dataset == !!dataset)
    choice_names <- purrr::map2(rows$Description, rows$Link, function(desc, link) {
      spec_link <- if (is.na(link)) "" else paste0(
        "<a href='", link, "' target='_blank' onclick='event.stopPropagation()' ",
        "class='tlg-spec' title='View spec'>", spec_icon_html, "</a>"
      )
      HTML(paste0("<span class='tlg-desc'>", htmltools::htmlEscape(desc), "</span>", spec_link))
    })
    div(
      class = "tlg-ds", `data-dataset` = dataset,
      checkboxGroupInput(
        inputId = ns(input_id),
        label = NULL,
        choiceNames = choice_names,
        choiceValues = as.character(rows$id)
      )
    )
  }

  # One flex column per Type (plain flex, not the bootstrap grid, whose negative
  # row margins would misalign the columns against the toolbar).
  type_columns <- purrr::map(present_types, function(tp) {
    tp_pairs <- dplyr::rename(pairs[pairs$Type == tp, ], type = Type, dataset = Dataset)
    div(
      class = "tlg-col",
      div(
        class = "tlg-col-head",
        tags$span(
          class = "tlg-col-title",
          icon(.TLG_TYPE_ICON[[tp]]), paste0(" ", tp, "s"),
          tags$span(sum(avail$Type == tp), class = "tlg-col-count")
        ),
        tags$button(
          type = "button", class = "tlg-col-selall",
          onclick = "window.tlgAdd.colSelect(this)", "Select all"
        )
      ),
      div(
        class = "tlg-col-body",
        purrr::pmap(tp_pairs, build_group_ui),
        div(class = "tlg-col-empty", "None in this view", style = "display: none;")
      )
    )
  })

  # Dataset tab bar; first dataset active by default.  data-total feeds the count
  # badge (restored when the search box is cleared).
  tab_bar <- div(
    class = "tlg-tabs",
    purrr::imap(datasets, function(ds, i) {
      ds_total <- sum(avail$Dataset == ds)
      tags$button(
        type = "button",
        class = paste("tlg-tab", if (i == 1) "active" else ""),
        `data-dataset` = ds, `data-total` = ds_total,
        onclick = paste0("window.tlgAdd.setTab(", tlg_js_str(ds), ", this)"),
        ds, tags$span(ds_total, class = "tlg-tab-count")
      )
    })
  )

  # Per-open initialisation: pick the first dataset tab and render.  Kept inline
  # (not in tlg_add_picker.js) because it depends on the datasets present now.
  init_js <- paste0(
    "window.tlgAdd.tab = ", tlg_js_str(datasets[1]), "; ",
    "window.tlgAdd.q = ''; window.tlgAdd.render();"
  )

  # Shared left inset so toolbar, tabs, column headers and checkbox rows all line
  # up on the same left edge (see --tlg-inset in _tlg_add_modal.scss).
  ui <- div(
    class = "tlg-add-modal",
    div(
      class = "tlg-toolbar",
      tags$input(
        type = "text", class = "form-control tlg-search-input",
        placeholder = "Search outputs…",
        oninput = "window.tlgAdd.setQuery(this.value)"
      ),
      tags$button(type = "button", class = "btn btn-sm btn-default",
                  onclick = "window.tlgAdd.selectAll()", "Select all"),
      tags$button(type = "button", class = "btn btn-sm btn-default",
                  onclick = "window.tlgAdd.clearAll()", "Clear all"),
      div(class = "tlg-toolbar-sep"),
      downloadButton(ns("modal_dl_csv"), "CSV", class = "btn-sm btn-default"),
      downloadButton(ns("modal_dl_xlsx"), "XLSX", class = "btn-sm btn-default")
    ),
    tab_bar,
    div(class = "tlg-add-checklist tlg-cols", type_columns),
    div(class = "tlg-no-matches", "No outputs match your search.", style = "display: none;"),
    tags$script(HTML(init_js))
  )

  list(ui = ui, group_ids = pairs$input_id)
}

#' Ids of the rows checked in the add-picker modal, mapped back to `tlg_order()`.
#'
#' @param input The module `input` object.
#' @param group_ids Character vector of `checkboxGroupInput` ids in the modal.
#' @returns Integer vector of checked `tlg_order()` ids (empty if none).
#' @noRd
checked_tlg_ids <- function(input, group_ids) {
  as.integer(unlist(lapply(group_ids, function(gid) input[[gid]])))
}

#' Available-TLG catalog for the modal's CSV / XLSX download.
#'
#' @param df The available-TLG tibble (`modal_avail()`), or `NULL`.
#' @returns A data frame with `Type`, `Dataset`, `PKid`, `Description`.
#' @noRd
tlg_modal_dl_data <- function(df) {
  if (is.null(df) || nrow(df) == 0) {
    return(data.frame(Type = character(), Dataset = character(),
                      PKid = character(), Description = character()))
  }
  dplyr::select(df, Type, Dataset, PKid, Description)
}
